cpp/ql/src/Likely Bugs/UnderspecifiedFunctions/MistypedFunctionArguments.c

```diff
@@ -0,0 +1,8 @@
+void no_argument();
+
+void three_arguments(int x, int y, int z);
+
+void calls() {
+	three_arguments(1, 2, 3); // GOOD
+	three_arguments(1.0f, 2, 3.0f); // BAD: arguments 1 and 3 do not match parameters
```

Comment:
I don't think this is `BAD`. The declaration of `three_arguments` has a prototype, so the floats will be converted to `int` semantically -- rounded down, not re-interpreted as a nonsense bit pattern.

Reply:
When compiling the following file:
<pre>
//void three_arguments();
void three_arguments(int x, char y, int *z);
int var;
double  d = 3732.70e13L;
extern unsigned long long *ull;

void calls() {
	three_arguments(1, 2, &var); // GOOD
	three_arguments(1U, 2U, &var);    // BAD: arguments 1, 2 and 3 do not match parameters
	three_arguments(1.0f, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(1.0L, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(d, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(2ULL, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(*ull, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(&ull, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(&var, 2, &var); // BAD: arguments 1 and 3 do not match parameters
	three_arguments(var, 2, 3.0f); // BAD: arguments 1 and 3 do not match parameters
}
</pre>

Both <code>gcc</code> and <code>clang</code> only issue a diagnostic on the last 3 invocations of <code>three_arguments</code>, and a warning at that.  This even with <code>--pedantic</code>.  All of these involve pointers.  

So the question is, which arguments shall we deem compatible when constructing <code>MistypedFunctionArguments.ql</code>?  As I've made clear in this PR thread, I believe that all but the first invocation of <code>three_arguments</code> are bad (and the query in its present state mostly reflects this), but it seems that my position is becoming more untenable by the day.  Should we just warn on pointer-related mismatches? What do people think?